src/Query/ConditionGroup.php

```diff
@@ -0,0 +1,194 @@
+ *   $condition_group = new ConditionGroup();
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition((new ConditionGroup('OR'))
+ *     ->addCondition('uid', 1)
+ *     ->addCondition('status', '1')
+ *   );
+ * @endcode
+ */
+final class ConditionGroup implements \Countable, CacheableDependencyInterface {
+
+  use RefinableCacheableDependencyTrait;
+
+  /**
+   * The conditions.
+   *
+   * @var \Drupal\entity\Query\Condition[]|\Drupal\entity\Query\ConditionGroup[]
+   */
+  protected $conditions = [];
+
+  /**
+   * The conjunction.
+   *
+   * @var string
+   */
+  protected $conjunction;
+
+  /**
+   * Constructs a new ConditionGroup object.
+   *
+   * @param string $conjunction
+   *   The conjunction.
+   */
+  public function __construct($conjunction = 'AND') {
+    $this->conjunction = $conjunction;
+  }
+
+  /**
+   * Gets the conjunction.
+   *
+   * @return string
+   *   The conjunction. Possible values: AND, OR.
+   */
+  public function getConjunction() {
+    return $this->conjunction;
+  }
+
+  /**
+   * Adds a condition.
+   *
+   * @param string|\Drupal\entity\Query\ConditionGroup $field
+   *   Either a condition group (for nested AND/OR conditions), or a
+   *   field name with an optional column name. E.g: 'uid', 'address.locality'.
+   * @param mixed $value
+   *   The value.
+   * @param string $operator
+   *   The operator.
+   *   Possible values: =, <>, <, <=, >, >=, BETWEEN, NOT BETWEEN,
+   *                   IN, NOT IN, IS NULL, IS NOT NULL.
+   *
+   * @return $this
+   */
+  public function addCondition($field, $value = NULL, $operator = NULL) {
+    if ($field instanceof ConditionGroup) {
+      if ($field->count() === 1) {
+        // The condition group only has a single condition, merge it.
+        $this->conditions[] = reset($field->getConditions());
+        $this->addCacheTags($field->getCacheTags());
+        $this->addCacheContexts($field->getCacheContexts());
+        $this->mergeCacheMaxAge($field->getCacheMaxAge());
+      }
+      elseif ($field->count() > 1) {
+        $this->conditions[] = $field;
+      }
+    }
+    else {
+      $this->conditions[] = new Condition($field, $value, $operator);
+    }
+
+    return $this;
+  }
+
+  /**
+   * Gets all conditions and nested condition groups.
+   *
+   * @return \Drupal\entity\Query\Condition[]|\Drupal\entity\Query\ConditionGroup[]
+   *   The conditions, where each one is either a Condition or a nested
+   *   ConditionGroup. Returned by reference, to allow callers to replace
+   *   or remove conditions.
+   */
+  public function &getConditions() {
+    return $this->conditions;
+  }
+
+  /**
+   * Clones the contained conditions when the condition group is cloned.
+   */
+  public function __clone() {
+    foreach ($this->conditions as $i => $condition) {
+      $this->conditions[$i] = clone $condition;
+    }
+  }
+
+  /**
+   * Gets the string representation of the condition group.
```

Comment:
Is this for development purposes also?

Reply:
It must be, because: `return (string) reset($this->conditions);`

So, maybe add that remark to this comment too?